File: grammar/c++10/rules.cpp
#include "rules.hpp"

#include <iostream>
using std::cout;
using std::endl;

template<>
parser::result::any parser::parse(symbol::not_delimeter, frame::text & frame, char next_char)
{
    cout << __PRETTY_FUNCTION__ << endl;
    if(isalnum(next_char) || next_char == ' ')
    {
        frame.current_text->content.push_back(next_char);
        return {result::finished{.consume = true, .frame = false}};
    }
    else
    {
        return {result::mismatch{}};
    }
}

template<>
parser::result::any parser::parse(symbol::not_delimeter_more, frame::text & frame, char next_char)
{
    cout << __PRETTY_FUNCTION__ << endl;
    if(isalnum(next_char) || next_char == ' ')
    {
        frame.current_text->content.push_back(next_char);
        return {result::consumed{}};
    }
    else
    {
        return {result::finished{.consume = false, .frame = true}};
    }
}

template<>
parser::result::any parser::parse(symbol::value, frame::line & frame, char next_char)
{
    cout << __PRETTY_FUNCTION__ << endl;
    return
    {
        result::change_frame
        {
            .frame = frame::text{&frame.current_line->value}
        }
    };
}

template<>
parser::result::any parser::parse(symbol::character symbol, frame::line & frame, char next_char)
{
    cout << __PRETTY_FUNCTION__ << endl;
    switch (symbol.ch)
    {
        case ':':
            return {result::finished{.consume = true, .frame = false}};
        case '\n':
            return {result::finished{.consume = true, .frame = true}};
        default:
            return {result::mismatch{}};
    }
}

template<>
parser::result::any parser::parse(symbol::key, frame::line & frame, char next_char)
{
    cout << __PRETTY_FUNCTION__ << endl;
    return
    {
        result::change_frame
        {
            .frame = frame::text{&frame.current_line->key}
        }
    };
}

template<>
parser::result::any parser::parse(symbol::line, frame::line & frame, char next_char)
{
    cout << __PRETTY_FUNCTION__ << endl;
    return
    {
        result::append_symbols
        {
            .symbols = {symbol::key{}, symbol::character{':'}, symbol::value{}, symbol::character{'\n'}}
        }
    };
}

template<>
parser::result::any parser::parse(symbol::line, frame::message & frame, char next_char)
{
    cout << __PRETTY_FUNCTION__ << endl;
    frame.message->lines.push_back({});
    return
    {
        result::change_frame
        {
            .frame = frame::line{&frame.message->lines.back()}
        }
    };
}

template<>
parser::result::any parser::parse(symbol::message, frame::message & frame, char next_char)
{
    cout << __PRETTY_FUNCTION__ << endl;
    return
    {
        result::append_symbols
        {
            .symbols = {symbol::line{}, symbol::message{}}
        }
    };
}

template<>
parser::result::any parser::parse(symbol::message, frame::message & frame)
{
    cout << __PRETTY_FUNCTION__ << endl;
    return {result::finished{.consume = false, .frame = false}};
}

File: grammar/rust/rules.rs
use frame;
use parser;
use symbol;

impl parser::ParseInput for (symbol::Any, frame::Any<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        match self
        {
            (symbol::Any::NotDelimeter(symbol), frame::Any::Text(frame)) => (symbol, frame).parse(ch),
            (symbol::Any::NotDelimeterMore(symbol), frame::Any::Text(frame)) => (symbol, frame).parse(ch),
            (symbol::Any::Value(symbol), frame::Any::Line(frame)) => (symbol, frame).parse(ch),
            (symbol::Any::Character(symbol), frame::Any::Line(frame)) => (symbol, frame).parse(ch),
            (symbol::Any::Key(symbol), frame::Any::Line(frame)) => (symbol, frame).parse(ch),
            (symbol::Any::Line(symbol), frame::Any::Line(frame)) => (symbol, frame).parse(ch),
            (symbol::Any::Line(symbol), frame::Any::Message(frame)) => (symbol, frame).parse(ch),
            _ => parser::Result::Mismatch
        }
    }
}

impl parser::ParseNoInput for (symbol::Any, frame::Any<'_>)
{
    fn parse(self) -> parser::Result
    {
        match self
        {
            (symbol::Any::Message(symbol), frame::Any::Message(frame)) => (symbol, frame).parse(),
            _ => parser::Result::Mismatch
        }
    }
}

impl parser::ParseInput for (symbol::NotDelimeter, frame::Text<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        let (symbol, frame) = self;
        if ch.is_alphanumeric() || ch == ' '
        {
            parser::Result::Finished { consume: true, frame: false }
        }
        else
        {
            parser::Result::Mismatch
        }
    }
}

impl parser::ParseInput for (symbol::NotDelimeterMore, frame::Text<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        let (symbol, frame) = self;
        if ch.is_alphanumeric() || ch == ' '
        {
            parser::Result::Consumed
        }
        else
        {
            parser::Result::Finished { consume: false, frame: true }
        }
    }
}

impl parser::ParseInput for (symbol::Value, frame::Line<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        parser::Result::ChangeFrame
    }
}

impl parser::ParseInput for (symbol::Character, frame::Line<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        let (symbol, frame) = self;
        match symbol.0
        {
            ':' => parser::Result::Finished { consume: true, frame: false },
            '\n' => parser::Result::Finished { consume: true, frame: true },
            _ => parser::Result::Mismatch
        }
    }
}

impl parser::ParseInput for (symbol::Key, frame::Line<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        parser::Result::ChangeFrame
    }
}

impl parser::ParseInput for (symbol::Line, frame::Line<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        parser::Result::AppendSymbols(vec![
            symbol::Any::Key(symbol::Key),
            symbol::Any::Character(symbol::Character(':')),
            symbol::Any::Value(symbol::Value),
            symbol::Any::Character(symbol::Character('\n'))
        ])
    }
}

impl parser::ParseInput for (symbol::Line, frame::Message<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        let (symbol, frame) = self;
        parser::Result::ChangeFrame
    }
}

impl parser::ParseInput for (symbol::Message, frame::Message<'_>)
{
    fn parse(self, ch: char) -> parser::Result
    {
        parser::Result::AppendSymbols(vec![
            symbol::Any::Line(symbol::Line),
            symbol::Any::Message(symbol::Message)
        ])
    }
}

impl parser::ParseNoInput for (symbol::Message, frame::Message<'_>)
{
    fn parse(self) -> parser::Result
    {
        parser::Result::Finished { consume: false, frame: false }
    }
}
